teps/0096-pipelines-v1-api.md

```diff
@@ -174,11 +182,106 @@ V1 work should prioritize stabilizing existing beta features that support the [u
+
+### Production Readiness
+
+- Implement [TEP-0036: Start measuring performance](https://github.com/tektoncd/community/blob/main/teps/0036-start-measuring-tekton-pipelines-performance.md).
+- Implement [TEP-0073: Simplify metrics](https://github.com/tektoncd/community/blob/main/teps/0073-simplify-metrics.md).
+  - [TEP-0006: Tekton metrics](./0006-tekton-metrics.md) is out of scope because it discusses adding more granular metrics. More granular metrics may be added later.
+- Ensure metrics are accurate and mark all metrics as "alpha" or "stable". 
+
+### Stability
+
+- Track bug fixes and small backwards incompatible changes outside of this TEP (e.g. in a Github project or milestones).
+- Implement conversion webhook for v1beta1 objects.
+
+#### CRD Stability Levels
+
+- The Task, TaskRun, Pipeline, and PipelineRun CRDs should be moved to the "stable" stability level.
+- Because Custom Task is mission critical to our extensibility use case, Run should be upgraded to beta stability.
+  - See [Custom Tasks Beta](https://github.com/tektoncd/pipeline/issues/4313) for more information on work required to upgrade Custom Tasks to beta.
+- Keep the ClusterTask CRD at beta stability.
+  - ClusterTask functionality will likely be replaced by [remote resolution](./0060-remote-resource-resolution.md), but there's no plan
+  to replace this functionality at a beta level of stability before releasing a V1 API.
+  - [Still under discussion]: We will decide before v1 whether to stabilize or remove ClusterTask pending community feedback on ClusterTask and remote resolution.
+  See [Where should we take ClusterTasks next?](https://github.com/tektoncd/pipeline/issues/4476) for more info.
+
+| CRD         | Current level | Proposed level |
+| ----------- | ------------- | -------------- |
+| Task        | beta          | stable         |
+| TaskRun     | beta          | stable         |
+| Pipeline    | beta          | stable         |
+| PipelineRun | beta          | stable         |
+| ClusterTask | beta          | beta           |
+| Run         | alpha         | beta           |
+
+#### API Changes
+
+- Fix [pain points](https://github.com/tektoncd/pipeline/issues/3792) associated with TaskRun and PipelineRun Status.
+- Consider replacing any Task or Pipeline fields that should support parameterization with Strings and performing our own validation.
+  - See [Handling parameter interpolation in fields not designed for it](https://github.com/tektoncd/pipeline/issues/1530) for more details.
+
+#### Deprecations
+
+- Deprecate `PipelineRun.Timeout` and [upgrade `PipelineRun.Timeouts` to beta](https://github.com/tektoncd/pipeline/issues/4460).
+- Remove `Pipeline.Tasks[].Conditions`. Conditions have been replaced by when expressions, which are already in beta.
+- Because PipelineResources are deprecated, remove the `Resources` fields of `Task`, `TaskRun`, `Pipeline`, and `PipelineRun`.
```

Comment:
They won't be available in `v1` but they will be available in `v1beta` still?
I think it makes sense, we'll have to define what the behaviour will be like when the default stored version is `v1` but we still serve the `v1beta`... but it doesn't have to be defined in here.
/cc @sbwsg @vdemeester

Reply:
That's right. I think our compatibility policy already covers this. We'll have to support v1beta1 for 9 months after releasing v1. (Also allowed by our compatibility policy) we could remove `Resources` in v1beta1 9 months after announcing their deprecation.

Reply:
Yeah, I was thinking about how to technically support that rather then the policy. When a user submits a `v1beta` resource, it will be stored as `v1` - because we can only store one version.

So either we just the `v1` incompatible parts, but then I would argue it doesn't make sense to serve `v1beta`. Alternatively we store the `v1beta` only parts somewhere (annotations or so) and reconstruct the `v1beta` resource when it's requested.

We had this discussion when we introduced `v1beta`, but I don't remember what was the conclusion (if there was one) @vdemeester @sbwsg

Reply:
I see, thanks for bringing this up. I'll have to look into the best way to do this.

Reply:
@afrittoli note that it doesn't necessarily need to be stored as `v1`. We can have a period of time where the stored version is `v1beta1` but we publish `v1` as well.

/cc @sbwsg on the annotation mechanisms, I am a bit rusty too on this.

Reply:
We serialize the `Finally` section of a `v1beta1` Pipeline into an annotation if a client requests the `v1alpha1` version. When a client submits a `v1alpha1` version of a Pipeline with that annotation present the process reverses and the `v1beta1` object is recreated with the `Finally` section in-tact: https://github.com/tektoncd/pipeline/blob/7a0a988cafaa64ff3e08969fa66f8521638d9b52/pkg/apis/pipeline/v1alpha1/pipeline_conversion.go#L142-L177

We don't do this for any other `v1beta1`-only fields. So, for example, when a `v1alpha1` client requests a `v1beta1` Pipeline with When Expressions, we simply drop them on the floor. If that client then re-applies the `v1alpha1` version the stored `v1beta1` version will be updated and the `when` expressions will be gone.

For a while we were [working on an approach](https://github.com/tektoncd/pipeline/pull/3854) where we take the entire `v1beta1` object and serialize it into an annotation when clients request `v1alpha1`, but it proved to be a complex and imperfect workaround.

Reply:
I think this discussion is great and also a bigger conversation than this specific PR, so I've opened tektoncd/pipeline#4546 to discuss further.